sound/soc/sof/intel/tgl.c

```diff
@@ -20,12 +20,18 @@ static const struct snd_sof_debugfs_map tgl_dsp_debugfs[] = {
 	{"dsp", HDA_DSP_BAR,  0, 0x10000, SOF_DEBUGFS_ACCESS_ALWAYS},
 };
 
+static int tgl_shutdown(struct snd_sof_dev *sdev)
+{
+	sdev->system_suspend_target = SOF_SUSPEND_S3;
+	return snd_sof_suspend(sdev->dev);
+}
+
 /* Tigerlake ops */
 const struct snd_sof_dsp_ops sof_tgl_ops = {
 	/* probe/remove/shutdown */
 	.probe		= hda_dsp_probe,
 	.remove		= hda_dsp_remove,
-	.shutdown	= hda_dsp_remove,
+	.shutdown	= tgl_shutdown,
```

Comment:
While this may not cause problems on all HDA platforms, I think we should apply this logic to all HDA platforms and define this as hda_dsp_shutdown.  snd-hda-intel calls stop_chip() on all generations in its shutdown() (essentially puts the controller to D3-like state)

Reply:
agree with Kai, this makes no sense to me, this should be common to all platforms, and we need to check what happens if the probe failed in the first place.

Reply:
@libinyang have you tested it on the TGL chrome device wth WoV enabled? Is D0i3 -> D3 a valid transition in our state machine?

Reply:
it is absolutely not!

Reply:
@libinyang I get the feeling that calling suspend() in shutdown is not the right thing to do here. We cannot guarantee the state machine transition will be valid in all cases.

I asked this previously too. We need to figure out what is missing in hda_dsp_remove() compared to hda_suspend(). Maybe it is something as simple as the controller reset and add it to hda_dsp_remove(). Can you please check?

Reply:
Hi @ranj063 @plbossart calling suspend() to stop the DMA trace and the DSP FW can make me more relax about the status that no DSP SRAM is accessing and DSP cores can be powered off safely, so it at least looks no harm to me.

For the state machine transition, yes, the PR add some change, but since once the .shutdown() is called we will achieve the Final State and the life cycle of the state machine is finished, it will start from new Initial State D0I0 at the subsequent reboot. So this looks reasonable?

```
/*
 * Audio DSP states may transform as below:-
 *                    +-------------+
 *                    |             |
 *                    |Initial State|
 *                    |             |
 *                    +-----+-------+
 *                          |boot
 *                          |              Opportunistic D0I3 in S0
 *     Runtime    +---------v-----------+  Delayed D0i3 work timeout
 *     suspend    |                     +--------------------+
 *   +------------+       D0I0(active)  |                    |
 *   |            |                     <---------------+    |
 *   |   +-------->                     |    New IPC	|    |
 *   |   |Runtime +--^--+---------^--+--+ (via mailbox)	|    |
 *   |   |resume     |  |         |  |			|    |
 *   |   |           |  |         |  |			|    |
 *   |   |     System|  |         |  |			|    |
 *   |   |     resume|  | S3/S0IX |  |                  |    |
 *   |   |	     |  | suspend/|  | S0IX             |    |
 *   |   |           |  | shutdown|  |suspend/          |    |
 *   |   |           |  |         |  |shutdown          |    |
 *   |   |           |  |         |  |                  |    |
 * +-v---+-----------+--v-------+ |  |           +------+----v----+
 * |                            | |  +----------->                |
 * |       D3 (suspended)       | |              |      D0I3      |
 * |                            | +--------------+                |
 * |                            |  System resume |                |
 * +----------------------------+		 +----------------+
 *               |
 *               |             +-----------+
 *               |  shutdown   |           |
 *               +------------>|Final State|
 *                             |           |
 *                             +-----------+
 *
 * S0IX suspend: The DSP is in D0I3 if any D0I3-compatible streams
 *		 ignored the suspend trigger. Otherwise the DSP
 *		 is in D3.
 */
```

Reply:
@keyonjie I don't follow your points, there is still no way to go from D0i3 to D3 to Final state.

Reply:
@plbossart @keyonjie yes, this change is good. The CTX_SAVE IPC in sof_suspend() will bring the DSP back to D0i0 before putting it into D3.

Reply:
I still don't get the flow, and btw why do we not release pci resources on shutdown?

````
void sof_pci_remove(struct pci_dev *pci)
{
	/* call sof helper for DSP hardware remove */
	snd_sof_device_remove(&pci->dev);

	/* follow recommendation in pci-driver.c to increment usage counter */
	if (snd_sof_device_probe_completed(&pci->dev) &&
	    !(sof_pci_debug & SOF_PCI_DISABLE_PM_RUNTIME))
		pm_runtime_get_noresume(&pci->dev);

	/* release pci regions and disable device */
	pci_release_regions(pci);
}
EXPORT_SYMBOL_NS(sof_pci_remove, SND_SOC_SOF_PCI_DEV);

void sof_pci_shutdown(struct pci_dev *pci)
{
	snd_sof_device_shutdown(&pci->dev);
}
EXPORT_SYMBOL_NS(sof_pci_shutdown, SND_SOC_SOF_PCI_DEV);
````

not sure why pci_release_regions() is missing nor why we don't test if the probe succeeded.

Reply:
@plbossart @ranj063 What do you think @keyonjie 's flow? Based on the flow, in the shutdown, the audio will be in the D0 and audio driver call suspend() to put it in D3 state. 

Based on the hardware datasheet, we should put audio in D3 when S4/S5.

Reply:
> While this may not cause problems on all HDA platforms, I think we should apply this logic to all HDA platforms and define this as hda_dsp_shutdown. snd-hda-intel calls stop_chip() on all generations in its shutdown() (essentially puts the controller to D3-like state)

@kv2019i Do you mean we should rename tgl_shutdown() to hda_dsp_shutdown() and put it in hda-dsp.c? And set .shutdown	= hda_dsp_shutdown, in the platform file such as tgl.c?

Reply:
@plbossart @ranj063  -- although a bit scarce on detail, the guidance in ACPI spec is to system in "similar" state as in S4. Actions like freeing PCI resources are not part of S4 sequence and Linux kernel is not expecting this either. I think @keyonjie you picked dsp_remove() as that did what was needed on the TGL case, but it's not like we have a _requirement_ to do all actions that are done at remove.

In general, not too many drivers implement a lot of logic in shutdown callbacks, but if you look at Intel examples, there is some reuse of the suspend() logic. One example is Intel ipw2x00 network driver where same ipw2100_down() is used to implement the majority of the logic for both driver suspend() as well as for shutdown().

I do think reusing suspend() makes sense. Based on ACPI specs, the requirements are for most part the same. But if we have a regression/problem in shutdown, it's a bit painful to debug as you need to hook up a serial console to the system and there will be nothing in the CI logs.

Reply:
@libinyang wrote:
> @kv2019i Do you mean we should rename tgl_shutdown() to hda_dsp_shutdown() and put it in hda-dsp.c? And set .shutdown = hda_dsp_shutdown, in the platform file such as tgl.c?

Ack, exactly that.

Reply:
I remained unconvinced by the changes at the HDaudio level, it's missing the detection of successful probe we see in sof_pci_remove() and the release of PCI resources.

Reply:
@kv2019i well, it's either the same as suspend or same as remove. it seems we can't figure out what the actual requirement is...

Reply:
@plbossart :) well that's exactly what I tried to clarify above. The spec only states S5 should be same as S4, so this is all about suspend and no linkage to remove. Unless anyone can point me to a spec, I'll be sticking to this.

Reply:
@kv2019i you may be right at the spec level, I don't see in the Linux code any reference to the parent-child model that's used for suspend. In theory if we followed the suspend model we would call .shutdown on the children before moving up the parents, but I don't see that code. When we used the aux bus for SoundWire, the use of .shutdown led to a device getting stuck, and we could never figure out why. So there is still some voodoo logic that we haven't understood somewhere in the kernel.

Reply:
@plbossart dsp_remove() is used in module removal. It's a software conception. We need to release pci resource in module remove. This is because we may reload the kernel modules without reboot. In this case, we must make sure these resource are not occupied. But for shutdown, we don't need to release these resource, because after reboot all the resource will automatically released. I think it may be OK to add the PCI resource release. But this action is not required from HW. HW team only requires driver to enter D3. I'm not sure PCI level's action when we release PCI resource. If PCI make some PCI clock or power switching when PCI resource release, this may be not acceptable by HW.